### PR TITLE
Refactor /ds to compile learnsets from prevos and certain baseSpecies

### PR DESCRIPTION
**Pull request by augmt**
This PR fixes [this][1] issue from the Bug Reports topic and employs @TheImmortal's suggested approach from #1877.

(Someone should attest that the refactor's execution speed is acceptable.)

  [1]: http://www.smogon.com/forums/threads/bug-reports-v2-0-read-op-before-posting.3469932/page-212#post-6214252

### CODE REVIEW

**Comment by Slayer95** [comment]
This effectively reverts https://github.com/Zarel/Pokemon-Showdown/commit/6f504d7c4fd84292b7064923df15b64d87056448
I don't think this is what TI referred to. Just adding a bit more of the `checkLearnset` logic could be enough.

(I'll try to run benchmarks, but I may not be able to do it fast enough)

Tagging @Zarel.

**Comment by gamut-was-taken** [comment]
What if I undo my changes and add to the while block in the `compileLearnsets` case, like so:

    while (template.baseSpecies !== template.species || template.prevo) {
    	if (template.baseSpecies !== template.species) {
    		template = Tools.getTemplate(template.baseSpecies)
    	} else {
    		template = Tools.getTemplate(template.prevo)
    	}
    	for (var move in template.learnset) {
    		if (!fullLearnset[move]) fullLearnset[move] = template.learnset[move];
    	}
    }

**Comment by SolarisFox** [comment]
`TeamValidator.checkLearnsetSync` is much slower (more than 5 times) in comparison to a simple check if the move exists within the Pokemon's learnset. Further, this isn't fully utilizing the full power of `TeamValidator.checkLearnsetSync` by ignoring the `format` parameter, which it could potentially be using with either tier or gen parameter information.

**Comment by gamut-was-taken** [comment]
How's this, then?

**Comment by Zarel** [comment]
Looks better but if we add the loop here, we should remove it from `checkLearnsets` and `checkLearnsetsSync`, right?

**Comment by Zarel** [comment]
also ick RAM usage

**Comment by gamut-was-taken** [comment]
I'm not sure I follow. This pull request no longer uses `checkLearnset` or `checkLearnsetSync` since they gobble up RAM, just as before.

File: chat-plugins/info.js [hunk 1]
```diff
@@ -531,8 +531,12 @@ var commands = exports.commands = {
...
-						while (template.prevo) {
-							template = Tools.getTemplate(template.prevo);
+						while (template.prevo || (template.eventPokemon && template.baseSpecies !== template.species)) {
+							if (template.prevo) {
+								template = Tools.getTemplate(template.prevo);
+							} else {
+								template = Tools.getTemplate(template.baseSpecies);
+							}
 							for (var move in template.learnset) {
 								if (!fullLearnset[move]) fullLearnset[move] = template.learnset[move];
```
Comment (Zarel) [inline chat-plugins/info.js]: We're overwriting the `.learnsets` value in the dex.

If we do that, we've made `checkLearnsets`'s algorithm invalid. So `checkLearnsets` needs to be rewritten.

Comment (Zarel) [inline chat-plugins/info.js]: Also this needs to be moved out of `chat-plugins/info.js` and into `tools.js` or something. Modifications to core data files should not be done by commands.

Comment (gamut-was-taken) [inline chat-plugins/info.js]: Could we deep clone the templates [here][1] and [here][2]? 

(At the risk of answering my own question, I imagine this gobbles up quite an unacceptable bit of memory.)

  [1]: https://github.com/gamut-was-taken/Pokemon-Showdown/blob/dexsearch/chat-plugins/info.js#L467
  [2]: https://github.com/gamut-was-taken/Pokemon-Showdown/blob/dexsearch/chat-plugins/info.js#L531

Comment (Zarel) [inline chat-plugins/info.js]: I don't see the problem with just compiling the learnsets in `tools.js` for everything to use, including `checkLearnsets`.

**Comment by gamut-was-taken** [comment]
@Zarel I'm sorry but this is starting to seem futile. 

This pull request, if it were to be merged, would fix an issue with `dexsearch` but the command would still be inaccurate. `/ds bodyslam, celebrate`'s output would (and currently does) include Pikachu and Raichu even though they can't have both moves concurrently. A more complex algorithm would then be needed to fix that issue, but we'd be moving closer to reverting 6f504d7, or trading speed for accuracy, which is unacceptable.

I'm thinking `dexsearch`'s functionality should be left in tact.

**Comment by Zarel** [comment]
We could run `checkLearnsetSync` on only on the pokemon still available after running it through the fast learnset searcher.

**Comment by Slayer95** [comment]
And only on the 10 Pokémon that would be displayed. We could then change the message to `and *about* X more.`

**Comment by SolarisFox** [comment]
Slayer's suggestion is great, but I'm concerned with what happens when it's used with the `all` parameter. At that point would it not have possible searches where it would have to run `checkLearnsetSync` on all the pokemon after it's already tried to filter them out once? Of course we could just have it use `checkLearnsetSync` whenever there are X pokemon or less to try to avoid this, but no matter how much tolerance we give there will be cases of inconsistent search results.

Another thing worth noting is sets don't need to be validated if there is only one search parameter. We could make the assumption at that point that the most Pokemon it would have to validate is somewhere around 100 (for searches that are actually trying to find anything meaningful; dumb searches like `/ds protect, toxic, all` could still return almost all Pokemon).

**Comment by Slayer95** [comment]
<strike>Since `all` isn't broadcast anyway, we could run it client-side</strike>

**Comment by gamut-was-taken** [comment]
Could we suggest the user to narrow their search or possibly even do away with the `all` parameter for moves?

As an example, `/ds nastyplot, nightmare` could return:

    Meowth, Persian, Drowzee, Hypno, Mew, Aipom, Slowking, Girafarig, Houndour, Houndoom, and
    about 14 more. Redo the search with 'all' as a parameter to show all results or add more
    parameters for more refined results.

**Comment by Zarel** [comment]
We shouldn't run it client-side. I've actually been thinking about moving more of `/data` server-side, so the client doesn't need to load as much data.

File: chat-plugins/info.js [hunk 2]
```diff
@@ -552,6 +556,25 @@ var commands = exports.commands = {
 							}
 						}
 					}
+					if (Object.count(searches[search], true) > 1) {
```
Comment (SolarisFox) [inline chat-plugins/info.js]: It might be better to define this as a function earlier so it can be called in `recover:` and `priority:` as well.

**Comment by gamut-was-taken** [comment]
Verifying learnsets now has its own case.

**Comment by Zarel** [comment]
I still think moving `compileLearnsets` outside of chat plugins is the highest priority here.

**Comment by gamut-was-taken** [comment]
Seems unsettling, but it's done: `compileLearnsets` is out of chat-plugins.

**Comment by Zarel** [comment]
You're right, I forgot about that pullreq.

**Comment by SolarisFox** [comment]
I wasn't implementing it correctly. Your solution is better so I was going to close it when this is accepted.

**Comment by ascriptmaster** [comment]
How is progress on this?

**Comment by Zarel** [comment]
We should probably merge this.

**Comment by ascriptmaster** [comment]
@gamut-was-taken can you fix the merge conflicts?

**Comment by Zarel** [comment]
I can fix them manually if not. I'll need to review it in depth either way.

**Comment by ascriptmaster** [comment]
I took a quick look at what the merge conflicts were, they should be pretty simple for him to fix. Mostly seems to be indentation changes as a result of another commit, plus one line from the `/randpoke` PR that is trivial to fix.

**Comment by gamut-was-taken** [comment]
Done; well I hope.

**Comment by The-Diabolic-Gift** [comment]
How about making it possible to dexsearch by egg group

**Comment by ascriptmaster** [comment]
There's still merge conflicts. Have you tried rebasing yet?

**Comment by Zarel** [comment]
He did rebase, I've just been putting off merging because I still haven't fully worked through the consequences. When I do, I can fix merge conflicts manually, anyway.

**Comment by Slayer95** [comment]
At least this doesn't seem to introduce outstanding illegalities, as per #2057

**Comment by augmt** [comment]
I've gone and updated the code ([view working diff][1]) on my machine and came up with this: if a move exists in a Pokemon's learnset as an egg move, we can skip validating it. That'll greatly boost performance at the cost of accuracy.

The table below highlights the output and performance of /ds as it is on master and both scenarios in milliseconds (neither scenario has a limit on calls to `checkLearnsetSync`).


|  | as-is<br>([results][2]) | w/ checkLearnset + excluding egg moves<br>([results][3]) | w/ checkLearnset  + including egg moves<br>([results][4]) |
|--------------------------|:-----------------:|:--------------------------------------------------------------:|:---------------------------------------------------------------:|
| /ds bodyslam, celebrate | 76.5 | 83.9 | 77.6xx |
| /ds nastyplot, nightmare | 77.0 | 73.6 | 401.0x |
| /ds block, amnesia | 76.7 | 74.9 | 483.9x |
| /ds headbutt, charm | 78.0 | 82.4 | 1575.2 |

[1]: http://pastebin.com/VDUP3NkW
[2]: https://cloud.githubusercontent.com/assets/9887589/12154470/a3cabc38-b48d-11e5-840c-9fa99dc99de3.png
[3]: https://cloud.githubusercontent.com/assets/9887589/12154486/c1742184-b48d-11e5-9a19-da68b3086bc8.png
[4]: https://cloud.githubusercontent.com/assets/9887589/12154492/d037b79e-b48d-11e5-89b3-027c221c23e6.png


Does the working diff seem good or should I just rebase and update based on what the code's looked like so far?

**Comment by Zarel** [comment]
After further reflection, I'm not sure compiling learnsets really offers all that much performance benefit.

On the other hand, using ES6 Maps/Sets might help.

**Comment by Zarel** [comment]
I'm fine with checkLearnsets, obviously excluding egg moves.

**Comment by Zarel** [comment]
Although in gen 6, including egg moves should have minimal performance impact... why is the difference so big?

**Comment by augmt** [comment]
Including egg moves doesn't have as minimal an impact as other moves; it has some impact and that adds up. The more Pokemon in the search, the larger the difference. (I think my computer is overstating the difference though.)

**Comment by Zarel** [comment]
I think it's because we check past gen eggs in case there's a past-gen-only move in the learnset.

Have you tried hacking it to only check present-gen eggs? Or to check all eggs like present-gen eggs?

**Comment by augmt** [comment]
Just a bit. If I pass `format: {requirePentagon: true}` within `lsetData` then that minimizes checkLearnset's performance impact; otherwise it's the same as I've reported.

I'll update this PR late tonight if all's well.

**Comment by Zarel** [comment]
We don't want to require pentagons, though...

**Comment by Zarel** [comment]
Specifically, maybe we can pass something like `fastCheck` which is treated similarly to `requirePentagon` but still checks past gens.

**Comment by augmt** [comment]
Sorry should've said more. I was thinking that we'd only pass `requirePentagon` for searches that include egg move and skip validating when egg moves are searched along with moves that can only be acquired from past gens. I believe /ds' performance is fine in other cases anyway.

**Comment by Zarel** [comment]
Yes, but the `fastCheck` approach seems better. The idea is to quickly check past gens without checking chainbreeding

**Comment by augmt** [comment]
[So like this, roughly?](http://pastebin.com/3pav76r6)

**Comment by Zarel** [comment]
Nearly exactly like that, except I would make it a flag of `lsetData` instead of yet another argument.

**Comment by augmt** [comment]
Updated, except for the title.

**Pull request merged**